File: rpn.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


short s[2<<16], sp;

inline void
push(short a) { s[sp++] = a; }

inline short
peek(void) { return s[sp-1]; }

inline short
pop(void) { return s[--sp]; }

#define EQ(s1, s2) (!strcmp(s1, s2))

int
main(void)
{
	char *line, *token, *endp = NULL;
	size_t len, n;
	while (1) {
		printf("RPN> ");
		getline(&line, &len, stdin);
		token = strtok(line, "\t\r\n ");
		do {
			     if (EQ(token, ".")) { printf("%d\n", pop()); }
			else if (EQ(token, "=")) { printf("%d\n", peek()); }
			else if (EQ(token, "+")) { push((pop() + pop())); }
			else if (EQ(token, "-")) { push((n = pop(), pop() - n)); }
			else if (EQ(token, "*")) { push((pop() * pop())); }
			else if (EQ(token, "/")) { push((n = pop(), pop() / n)); }
			else if (EQ(token, "?")) {
				printf("Stack Pointer: %d\nStack: ", sp);
				for(int i = 0; i < sp; i++)
					printf("%d ", s[i]);
				printf("\n");
			} else {
				n = strtol(token, &endp, 10);
				if (endp && !(*endp))
					push((short) n);
			}
		} while ((token = strtok(NULL, "\t\r\n ")));
	}

	return 0;
}

File: rpn.ml
(* 64Kx16: word addressable 16-bit arch, like a typical Forth *)
let s = Bytes.create (2 lsl 16)

let sp = ref 0

let push n = Bytes.set_int16_ne s !sp n; sp := !sp + 2
let peek _ = Bytes.get_int16_ne s (!sp-2)
let pop  _ = let out = peek() in
    sp := !sp - 2; Bytes.set_int16_ne s !sp 0;
    out

let ws = Str.regexp "[\t\r\n ]+"
let nl = Str.regexp "^[+-]?[0-9]+$"

let printf = Printf.printf

let rec loop () =
    print_string "RPN> "
    |> read_line
    |> Str.split ws
    |> List.iter (function
         | "+" -> binop ( + )
         | "-" -> binop ( - )
         | "*" -> binop ( * )
         | "/" -> binop ( / )
         | "." -> printf "%d\n%!" (pop ())
         | "=" -> printf "%d\n%!" (peek())
         | "?" -> printf "Stack Pointer: %d\nStack: %S\n%!"
                         !sp (Bytes.sub_string s 0 32)
         | s when Str.string_match nl s 0 -> push (int_of_string s)
         | any -> ()
       )
    |> loop
and binop f = let n = pop () in
    push @@ f (pop()) n

let () = loop ()
